src/compute/block.rs

```diff
@@ -0,0 +1,655 @@
+//! Computes the [flexbox](https://css-tricks.com/snippets/css/a-guide-to-flexbox/) layout algorithm on [`Taffy`](crate::Taffy) according to the [spec](https://www.w3.org/TR/css-flexbox-1/)
+use crate::compute::LayoutAlgorithm;
+use crate::geometry::{Line, Point, Rect, Size};
+use crate::style::{AvailableSpace, Display, LengthPercentageAuto, Overflow, Position};
+use crate::style_helpers::TaffyMaxContent;
+use crate::tree::{CollapsibleMarginSet, Layout, RunMode, SizeBaselinesAndMargins, SizingMode};
+use crate::tree::{LayoutTree, NodeId};
+use crate::util::sys::f32_max;
+use crate::util::sys::Vec;
+use crate::util::MaybeMath;
+use crate::util::{MaybeResolve, ResolveOrZero};
+
+#[cfg(feature = "debug")]
+use crate::util::debug::NODE_LOGGER;
+
+/// The public interface to Taffy's Flexbox algorithm implementation
+pub struct BlockAlgorithm;
+impl LayoutAlgorithm for BlockAlgorithm {
+    const NAME: &'static str = "FLEXBOX";
+
+    fn perform_layout(
+        tree: &mut impl LayoutTree,
+        node: NodeId,
+        known_dimensions: Size<Option<f32>>,
+        parent_size: Size<Option<f32>>,
+        available_space: Size<AvailableSpace>,
+        _sizing_mode: SizingMode,
+        vertical_margins_are_collapsible: Line<bool>,
+    ) -> SizeBaselinesAndMargins {
+        compute(
+            tree,
+            node,
+            known_dimensions,
+            parent_size,
+            available_space,
+            RunMode::PeformLayout,
+            vertical_margins_are_collapsible,
+        )
+    }
+
+    fn measure_size(
+        tree: &mut impl LayoutTree,
+        node: NodeId,
+        known_dimensions: Size<Option<f32>>,
+        parent_size: Size<Option<f32>>,
+        available_space: Size<AvailableSpace>,
+        _sizing_mode: SizingMode,
+        vertical_margins_are_collapsible: Line<bool>,
+    ) -> Size<f32> {
+        compute(
+            tree,
+            node,
+            known_dimensions,
+            parent_size,
+            available_space,
+            RunMode::ComputeSize,
+            vertical_margins_are_collapsible,
+        )
+        .size
+    }
+}
+
+/// The intermediate results of a flexbox calculation for a single item
+struct BlockItem {
+    /// The identifier for the associated node
+    node_id: NodeId,
+
+    /// The order of the item
```

Comment:
What exactly does this mean?
Is this the index of the node in the parent item?

I think we can improve the comment here a bit to make it easier to understand for people who are not familiar with the spec

Reply:
> Is this the index of the node in the parent item?

Currently it is, yes. CSS calls this the "source order". Although CSS also supports an `order` property to override this, and it's kinda ambiguous as to whether this would end up being modified by that if Taffy supported it.